### ISSUE TITLE
Expose configuration option for changing the size of the buffer used to calculate checksums

### ISSUE THREAD

**Issue by korydraughn** [irods/irods#7947]
- [x] main
- [x] 4-3-stable
- [x] irods_docs (main only - *does not have 4-3-stable branch*)
- [x] S3 resource plugin docs (main)
- [x] S3 resource plugin docs (4-3-stable)
---
## Enhancement

The following is used as the buffer size for calculating checksums.

https://github.com/irods/irods/blob/ecaf66d06478bbe2a3e7be47d98f04ed97b94e9a/server/api/src/rsFileChksum.cpp#L21

Notice L342 in the following code snippet. The use of a 1MB buffer can result in a high number of read operations to the resource. iRODS needs to either increase the size of the buffer to a reasonable amount or expose a new configuration option that allows admins to set the value themselves.

https://github.com/irods/irods/blob/ecaf66d06478bbe2a3e7be47d98f04ed97b94e9a/server/api/src/rsFileChksum.cpp#L342-L359

Putting the configuration option in R_GRID_CONFIGURATION seems appropriate. And if necessary, the grid configuration option can be overridden via a local server_config.json option.

**Comment by trel** [comment]
And it's been that way since 2007(!)...

https://github.com/irods/irods-legacy/blame/ff4eaa47a34f1bb5990d5560f825975c26bab118/iRODS/server/api/src/rsFileChksum.c#L14

**Comment by trel** [comment]
What seems like a better default than 1MiB?   What is a good test to find a balance?    Just timing a 2GiB file with various buffer sizes?

**Comment by korydraughn** [comment]
I think we test a range of file sizes to see how different buffer/file sizes affect performance. Perhaps we try powers of 2 starting from 2<sup>0</sup> up to 2<sup>5</sup>. 2<sup>0</sup> is the baseline since that's what iRODS uses today.

As for file sizes, we try different sizes as well. We could test 500MiB, 1GiB, 2.5GiB, and 5GiB.

**Comment by trel** [comment]
Perhaps we even allow different buffer sizes for different file sizes...  Although that feels a bit much on first glance.

**Comment by korydraughn** [comment]
I agree. Benchmarking the operation in a somewhat realistic environment will show us the way.

**Comment by JustinKyleJames** [comment]
I see two different possible solutions to this.

1. Make that buffer size configurable.  A larger buffer size would reduce the number of requests.  This would be slower to implement but is the easiest fix.  We could even provide compiled version just for Carsten.

2. Add retries with some type of backoff when we get this error would essentially throttle how many requests iRODS would send as it would be waiting for the response from prior requests.  This could be done in the S3 plugin.

**Comment by korydraughn** [comment]
Option 1 is a must.

Option 2 is worth investigating as it could prove useful for cases where the buffer size option doesn't completely solve the problem.

**Comment by alanking** [comment]
For tracking things later...

https://github.com/irods/irods/pull/7964 implements the new feature - this can go in as soon as it's ready.
https://github.com/irods/irods_docs/pull/285 documents the new feature - this is being held until closer to 4.3.4 release.
https://github.com/irods/irods_resource_plugin_s3/pull/2217 documents using the new feature with S3 resource plugin - this is being held until closer to 4.3.4 release.

I've added checkboxes for the on-hold PRs so we don't prematurely close the issue and lose track of them.

**Comment by korydraughn** [comment]
irods/irods_docs#309 is merged _(supersedes irods/irods_docs#285 - forgot this PR existed)_.
irods/irods_resource_plugin_s3#2217 is merged.

The work for this issue is now complete.
